### PR TITLE
feat: embed Rekor bundle in OCI attestation layer annotations

### PR DESCRIPTION
**Pull request by ab-ghosh**
# Changes

This PR adds support for embedding the Rekor transparency log bundle (`dev.sigstore.cosign/bundle`) in OCI attestation layer annotations when `transparency.enabled=true`. This enables offline signature verification of attestations signed by Tekton Chains using keyless (Fulcio) signing, bringing Chains' attestation format in line with what `cosign attest` produces.

Fixes #1598

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [ ] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

``` release-note
Embed Rekor bundle in OCI attestation layer annotations to enable offline verification of keyless-signed attestations.
```

### CODE REVIEW

File: pkg/chains/signing/iface.go [hunk 1]
```diff
@@ -41,4 +42,6 @@ type Bundle struct {
 	Cert []byte
 	// Cert is an optional PEM encoded x509 certificate chain, if one was used for signing.
```
Comment (infernus01) [inline pkg/chains/signing/iface.go]: Not related to this PR but I just got eyes on this.
```suggestion
	// Chain is an optional PEM encoded x509 certificate chain, if one was used for signing.
```

**Review by infernus01 — CHANGES_REQUESTED** [review]
@ab-ghosh Would it be good to write this feature in the docs as well?

**Review by ab-ghosh — COMMENTED** [review]
@infernus01 Good eyes on the comment typo And yes, documenting this feature makes sense, let me add that.

File: pkg/chains/signing.go [hunk 2]
```diff
@@ -186,6 +187,32 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 			}
 			measureMetrics(ctx, metrics.SignedMessagesCount, o.Recorder)
 
+			// Upload to Rekor before storage so the bundle is available for OCI attestation annotations.
+			var rekorBundle *cbundle.RekorBundle
+			if shouldUploadTlog(cfg, tektonObj) {
+				rekorClient, err := getRekor(cfg.Transparency.URL)
+				if err != nil {
+					return err
+				}
```
Comment (infernus01) [inline pkg/chains/signing.go]: This seems to be a hard return. What if getRekor() fails (maybe due to bad rekor URL config or anything) and this might prevent all attestation storage. WDYT?

File: pkg/chains/storage/oci/legacy.go
Comment (infernus01) [inline pkg/chains/storage/oci/legacy.go:170]: One thing i noticed that the RekorBundle gets passed through in uploadAttestation below but not in uploadSignature here. since simplesigning + x509/fulcio + rekor is a valid config, the same offline-verification gap might exist for OCI image signatures too. right?
was this intentionally out of scope for this PR, or worth handling here as well?

File: pkg/chains/signing.go (hunk 2 above)
Comment (ab-ghosh) [reply pkg/chains/signing.go]: this hard return is pre-existing, IMO if getRekor fails, it's a config error, should not be treated as a transient failure

File: pkg/chains/storage/oci/legacy.go
Comment (ab-ghosh) [reply pkg/chains/storage/oci/legacy.go:170]: Yeah that totally makes sense, but #1598 is more aligned with attestations. We can address this in a separate issue/PR, what say?

Comment (infernus01) [reply pkg/chains/storage/oci/legacy.go:170]: Alright 👍🏻

**Review by infernus01 — COMMENTED** [review]
/lgtm

File: pkg/chains/signing.go [hunk 3]
```diff
@@ -186,6 +187,32 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
...
+
+				entry, err := rekorClient.UploadTlog(ctx, signer, signature, rawPayload, signer.Cert(), string(payloadFormat))
+				if err != nil {
+					logger.Warnf("error uploading entry to tlog: %v", err)
```
Comment (jkhelil) [inline pkg/chains/signing.go:209]: PLease document this behaviour, if an error occur, it doesnt block storage but bundle will be unavailable." Consumers that rely on the bundle annotation being present will silently get an attestation without a bundle.

File: pkg/chains/signing.go [hunk 4]
```diff
@@ -186,6 +187,32 @@ func (o *ObjectSigner) Sign(ctx context.Context, tektonObj objects.TektonObject)
 			}
 			measureMetrics(ctx, metrics.SignedMessagesCount, o.Recorder)
 
+			// Upload to Rekor before storage so the bundle is available for OCI attestation annotations.
+			var rekorBundle *cbundle.RekorBundle
+			if shouldUploadTlog(cfg, tektonObj) {
+				rekorClient, err := getRekor(cfg.Transparency.URL)
```
Comment (jkhelil) [inline pkg/chains/signing.go]: getRekor(cfg.Transparency.URL) constructs a brand-new Rekor HTTP client on every iteration of the for _, obj := range objects loop.
The client is stateless and config-driven, so it can be safely lifted to before the objects loop.

File: pkg/chains/signing_test.go [hunk 5]
```diff
@@ -573,13 +719,22 @@ type mockRekor struct {
 func (r *mockRekor) UploadTlog(ctx context.Context, signer signing.Signer, signature, rawPayload []byte, cert, payloadFormat string) (*models.LogEntryAnon, error) {
```
Comment (jkhelil) [inline pkg/chains/signing_test.go:775]: The code path at signing.go:209 ("Rekor entry missing verification data, skipping bundle") is never exercised. Add a test or a second mock variant where Verification is nil, asserting that RekorBundle is nil in StorageOpts even when tlog upload succeeds.

File: docs/sigstore.md [hunk 6]
```diff
@@ -35,6 +35,12 @@ Right now, Chains default to storing entries in the public Rekor instance
 kubectl patch configmap chains-config -n tekton-chains -p='{"data":{"transparency.url": "<YOUR URL>"}}'
 ```
 
+### Offline Verification with Rekor Bundle
+
+When transparency is enabled and the OCI storage backend is used, Chains automatically embeds the Rekor bundle (`dev.sigstore.cosign/bundle`) in the OCI attestation layer annotations. This enables offline verification of transparency log entries without querying the Rekor server.
+
+No additional configuration is required — the bundle is embedded automatically when `transparency.enabled` is set to `true` and attestations are stored in an OCI registry.
```
Comment (jkhelil) [inline docs/sigstore.md:44]: Add a sentence that this applies to attestations only (in-toto / DSSE-wrapped SLSA); simplesigning (OCI image signature via simplesigning format) does not currently embed the bundle. Without this, users may set up Fulcio+Rekor+OCI storage, observe that cosign verify (not verify-attestation) doesn't find the bundle, and be confused.

**Comment by ab-ghosh** [comment]
Thank you for the review, @jkhelil! I've addressed all your suggestions:
  - Documented the behavior when Rekor upload fails
  - Lifted getRekor() out of the objects loop
  - Added TestSigner_NilVerificationSkipsBundle for the nil Verification path
  - Clarified in docs that the bundle applies to attestations only
 
  Please take a look when you get a chance.

**Comment by jkhelil** [comment]
@anithapriyanatarajan can you have a look please

File: docs/sigstore.md (hunk 6 above)
Comment (anithapriyanatarajan) [reply docs/sigstore.md:44]: @ab-ghosh - is this done?

File: pkg/chains/signing/iface.go [hunk 7]
```diff
@@ -39,6 +40,8 @@ type Bundle struct {
 	Signature []byte
 	// Cert is an optional PEM encoded x509 certificate, if one was used for signing.
 	Cert []byte
-	// Cert is an optional PEM encoded x509 certificate chain, if one was used for signing.
+	// Chain is an optional PEM encoded x509 certificate chain, if one was used for signing.
```
Comment (anithapriyanatarajan) [inline pkg/chains/signing/iface.go:43]: ❤️

**Comment by anithapriyanatarajan** [comment]
@ab-ghosh Thank you for this PR. Verified that with PR rekor bundle for reference
`{
  "schemaVersion": 2,
  "mediaType": "application/vnd.oci.image.manifest.v1+json",
  "config": {
    "mediaType": "application/vnd.oci.image.config.v1+json",
    "size": 233,
    "digest": "sha256:f772b2a04aa836eba6af79e20228f0364fb0fa0877038884c56d5524f3880846"
  },
  "layers": [
    {
      "mediaType": "application/vnd.dsse.envelope.v1+json",
      "size": 2654,
      "digest": "sha256:cab0e1e0dc66c9084883ced14fb5b5d4d62e5f922c3b9e7fcedbefde38cc36aa",
      "annotations": {
        "dev.cosignproject.cosign/signature": "",
        **"dev.sigstore.cosign/bundle": "{\"SignedEntryTimestamp\":\"MEYCIQCXuOiPPOVB6ZlVHmQc9lc2JPKSwOOjbgW+KOEm90P5FAIhALERFX+qw2ocJBR1z5ucs+9711QUCIiStdXTZD9CsM4z\",\"Payload\":{\"body\":\"eyJhcGlWZXJzaW9uIjoiMC4wLjEiLCJraW5kIjoiaW50b3RvIiwic3BlYyI6eyJjb250ZW50Ijp7Imhhc2giOnsiYWxnb3JpdGhtIjoic2hhMjU2IiwidmFsdWUiOiJjYWIwZTFlMGRjNjZjOTA4NDg4M2NlZDE0ZmI1YjVkNGQ2MmU1ZjkyMmMzYjllN2ZjZWRiZWZkZTM4Y2MzNmFhIn0sInBheWxvYWRIYXNoIjp7ImFsZ29yaXRobSI6InNoYTI1NiIsInZhbHVlIjoiY2M3MTk5MTE3NmZjZDAyNjlkMmI3OTNlNDE3MDYwNTcyNWIzMDM3MjE5ZDI1ODEzNDRjM2RlOWMzMzEzN2YyMSJ9fSwicHVibGljS2V5IjoiTFMwdExTMUNSVWRKVGlCUVZVSk1TVU1nUzBWWkxTMHRMUzBLVFVacmQwVjNXVWhMYjFwSmVtb3dRMEZSV1VsTGIxcEplbW93UkVGUlkwUlJaMEZGV1ZGd1pqUlFSRE5LWmxOTGIzRkNPV2xEUTNWSVNtcGxUMnRMVVFvMmJXWlpjelV5VEd4eGR6ZFRjMG92Y2t0UmJWTXZZVkJSV2tOVk9VVnFZMlpQUzA1dFZETXhTalkxU1VjemJGbEVhMFZZTjA5c05VVjNQVDBLTFMwdExTMUZUa1FnVUZWQ1RFbERJRXRGV1MwdExTMHRDZz09In19\",\"integratedTime\":1783071679,\"logIndex\":2058342836,\"logID\":\"c0d23d6ad406973f9559f3ba2d1ca01f84147d8ffc5b8445c224f98b9591801d\"}}",**
        "dev.sigstore.cosign/certificate": "",
        "dev.sigstore.cosign/chain": ""
      }
    }
  ]
}
`
Happy to lgtm. If possible please post a blog with cosign steps to verify. Thank you.
